In [11]:
from bloom_filter import BloomFilter
import json
import datetime
import sys
import random

def url_test(positives, negatives, fp_rate):
    bloom_capacity = len(positives)
    bf = BloomFilter(bloom_capacity, fp_rate)
    count = 0

    before = datetime.datetime.now()
    for pos in positives:
        if count == bloom_capacity:
            break
        bf.add(pos)
        count += 1
        assert(pos in bf)
    
    after = datetime.datetime.now()
    delta = after - before
    print("Build time:", delta)
    print("Number of bits:", bf.num_bits_m)
#     print("Converting bits to bytes", (bf.size)/8)
#     print("Hash functions needed", bf.hash_count)

    fp = 0.0
    for neg in negatives:
        if(str(neg) in bf):
            fp += 1
    print("Results of url test:")
    print("False positives", fp / len(negatives))


with open('dataset.json', 'r') as f:
    dataset = json.load(f)

positives = dataset['positives']
negatives = dataset['negatives']
data_fraction = 1
positives = positives[:int(data_fraction * len(positives))]
negatives = negatives[:int(data_fraction * len(negatives))]

print("Total Dataset: ", len(positives) + len(negatives))
print("Number of positive instances: ", len(positives))
print("Number of negative instances: ", len(negatives))

url_test(positives, negatives, 0.01)

Total Dataset:  2926705
Number of positive instances:  1491178
Number of negative instances:  1435527
Build time: 0:01:09.696275
Number of bits: 14293029
Results of url test:
False positives 0.010015833906293646


#### Results of the Bloom Filter on Half Dataset

Total Dataset:  1463352  
Number of positive instances:  745589  
Number of negative instances:  717763  
Build time: 0:00:33.993262  
Number of bits: 7146515    
Results of url test:  
False positives 0.009845868343729058  
Size in bytes: 893314.375   
Size in Mb: 0.893314375  

#### Results of the Bloom Filter Using Full Dataset

Total Dataset:  2926705  
Number of positive instances:  1491178  
Number of negative instances:  1435527  
Build time: 0:01:09.696275  
Number of bits: 14293029  
Results of url test:  
False positives 0.010015833906293646  
Size in bytes: 1786628.625  
Size in Mb: 1.786628625  